proposals/NNNN-isolated-default-values.md

```diff
@@ -0,0 +1,173 @@
+
+```swift
+@MainActor func requiresMainActor() -> Int { ... }
+
+@MainActor func useDefault(value: Int = requiresMainActor()) { ... }
+```
+
+In the above code, the default argument for `value` requires `@MainActor` isolation, because the default value calls `requiresMainActor()` which is isolated to `@MainActor`.
+
+A default value expression must only have one required isolation; it is an error for a default value expression to contain multiple callees with different actor isolation. For example:
+
+```swift
+@MainActor func requiresMainActor() -> Int { ... }
+@AnotherActor func requiresAnotherActor() -> Int { ... }
+
+@MainActor func useDefault(
+  value: (Int, Int) = (requiresMainActor(), requiresAnotherActor()) // error!
+) {}
+```
+
+The above example is invalid because the default argument for `value` requires both `@MainActor` and `@AnotherActor`, but the caller can never satisfy both isolation requirements simultaneously.
+
+#### Closures
+
+Evaluating a closure literal itself can happen in any isolation domain; the actor isolation of a closure only applies when calling the closure. An actor-isolated closure enables the closure body to make calls within that isolation domain synchronously. For a closure literal in a default value expression that is not explicitly annotated with actor isolation, the inferred isolation of the closure is the union of isolation contexts of all callees in the closure body for synchronous calls. For example:
+
+```swift
+@MainActor func requiresMainActor() -> Int { ... }
+
+@MainActor func useDefaultClosure(
+  closure: () -> Void = {
+    requiresMainActor()
+  }
+) {}
+```
+
+The above `useDefaultClosure` function has a default argument value that is a closure literal. The closure body calls a `@MainActor`-isolated function synchronously, therefore the closure itself must be `@MainActor` isolated.
+
+Note that the inferred actor isolation of a closure literal can never be an actor instance based on the following two properties:
+
+1. To be isolated to an actor instance, a closure must either have its own (explicit) isolated parameter or capture an isolated parameter from its enclosing context.
+2. Closure literals in default arguments cannot capture values.
+
+### Enforcing default value isolation requirements
+
+#### Default argument values
+
+Isolation requirements for default argument expressions are enforced at the caller. If the caller is not in the required isolation domain, the default argument cannot be used and the argument must be specified explicitly. For example:
+
+```swift
+@MainActor func requiresMainActor() -> Int { ... }
+
+@MainActor func useDefault(value: Int = requiresMainActor()) { ... }
+
+@MainActor func mainActorCaller() {
+  useDefault() // okay
+}
+
+func nonisolatedCaller() async {
+  useDefault() // error
+
+  useDefault(value: await requiresMainActor()) // okay
+}
```

Comment:
```suggestion
func nonisolatedCaller() async {
  useDefault() // error

  useDefault(value: await requiresMainActor()) // okay
}
```
We'd probably expect the first line here to be written as:
```
await useDefault()
```
since we're moving to the main actor, no? And this second line would still need to `await` the `useDefault` call:
```
await useDefault(value: requiresMainActor())
```
We'd still be able to apply the rule that `await` can swallow multiple suspension points inside the expression, right?

Reply:
Ah yes, you're right for both cases, thank you!